codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocol.kt

```diff
@@ -0,0 +1,206 @@
+import software.amazon.smithy.model.shapes.ResourceShape
+import software.amazon.smithy.model.shapes.ServiceShape
+import software.amazon.smithy.rust.codegen.rustlang.Writable
+import software.amazon.smithy.rust.codegen.rustlang.asType
+import software.amazon.smithy.rust.codegen.rustlang.rustTemplate
+import software.amazon.smithy.rust.codegen.rustlang.writable
+import software.amazon.smithy.rust.codegen.server.smithy.ServerCargoDependency
+import software.amazon.smithy.rust.codegen.smithy.CoreCodegenContext
+import software.amazon.smithy.rust.codegen.smithy.RuntimeType
+import software.amazon.smithy.rust.codegen.smithy.protocols.AwsJson
+import software.amazon.smithy.rust.codegen.smithy.protocols.AwsJsonVersion
+import software.amazon.smithy.rust.codegen.smithy.protocols.Protocol
+import software.amazon.smithy.rust.codegen.smithy.protocols.RestJson
+import software.amazon.smithy.rust.codegen.smithy.protocols.RestXml
+import software.amazon.smithy.rust.codegen.util.PANIC
+import software.amazon.smithy.rust.codegen.util.dq
+import software.amazon.smithy.rust.codegen.util.orNull
+
+fun allOperationShapes(service: ServiceShape, model: Model): List<OperationShape> {
+    val resourceOperationShapes = service
+        .resources
+        .mapNotNull { model.getShape(it).orNull() }
+        .mapNotNull { it as? ResourceShape }
+        .flatMap { it.allOperations }
+        .mapNotNull { model.getShape(it).orNull() }
+        .mapNotNull { it as? OperationShape }
+    val operationShapes = service.operations.mapNotNull { model.getShape(it).orNull() }.mapNotNull { it as? OperationShape }
+    return resourceOperationShapes + operationShapes
+}
+
+interface ServerProtocol {
+    // Returns the core `Protocol`.
+    fun coreProtocol(): Protocol
+
+    // Returns the Rust marker struct enjoying `OperationShape`.
+    fun markerStruct(): RuntimeType
+
+    // Returns the Rust router type.
+    fun routerType(): RuntimeType
+
+    // Returns the construction of the `routerType` given a `ServiceShape`, a collection of operation values
+    // (`self.operation_name`, ...), and the `Model`.
+    // TODO(Decouple): Perhaps this should lean on a Rust interface.
+    fun routerConstruction(service: ServiceShape, operationValues: Iterable<Writable>, model: Model): Writable
+
+    companion object {
+        // Upgrades the core protocol to a `ServerProtocol`.
+        fun fromCoreProtocol(coreCodegenContext: CoreCodegenContext, protocol: Protocol): ServerProtocol {
+            val serverProtocol = when (protocol) {
+                is AwsJson -> ServerAwsJsonProtocol(coreCodegenContext, protocol)
+                is RestJson -> ServerRestJsonProtocol(coreCodegenContext, protocol)
+                is RestXml -> ServerRestXmlProtocol(coreCodegenContext, protocol)
+                else -> throw IllegalStateException("unsupported protocol")
+            }
+            return serverProtocol
+        }
+    }
+}
+
+class ServerAwsJsonProtocol(
```

Comment:
So are these meant to replace the existing server protocol classes? Like, the `ServerAwsJson` class in `ServerAwsJson.kt` etc.

Reply:
I prefer this composition approach over the inheritance one we used over there though. Can you them into these ones, and split in several files?

And the factories should now return these ones e.g. `ServerRestJsonFactory`'s `protocol` method should return `ServerRestJson` etc...

Reply:
I'm not sure how to best merge them since inheritance has the advantage that we just override the method of the `Protocol` class that needs to return a different generator or that is subtly difference, and everything works correctly.

Maybe we need to use a combination of both: the truly server-specific methods go into this `ServerProtocol` interface, but then the classes that implement this interface should also inherit from `Protocol` to override things like e.g. `structuredDataSerializer`?

Reply:
Will track this effort/investigation here https://github.com/awslabs/smithy-rs/issues/1698